[Explorer] Add Metadata page

matextrem opened the pull request:
# Summary

Closes #160 

Added a new Metadata page where you can create a new app data document and also upload it to `IPFS`

<img width="886" alt="Screen Shot 2022-07-20 at 17 43 34" src="https://user-images.githubusercontent.com/11525018/180078920-2e33babc-b23e-44fa-a61f-dc5355dd5662.png">


# To Test

1. Open the page `metadata`
* You can complete the form and then hit `GENERATE METADATA DOC` button
* After the app data generation, you can also upload it to IPF by clicking  `UPLOAD APP DATA TO IPFS` button

matextrem commented:
@ramirotw could you try build the `lock.file` and push it ? I'm trying to do it locally but it's failing on my side.

ramirotw commented on src/apps/explorer/pages/Metadata/config.ts:
```diff
@@ -0,0 +1,60 @@
+      },
+    },
+  },
+}
```
shouldn't this be loaded through the sdk or the app-data package directly? @alfetopito I believe the json schema should be exposed through the sdk right? In that case maybe @matextrem can work locally with the app-data changes on the sdk before you publish version 1.0.0, wdyt?

alfetopito replied:
The SDK changes in regards to app-data are still pending review https://github.com/cowprotocol/cow-sdk/pull/50

Still, it is not exposing the appData schemas directly, only using it internally.
Do you think it makes sense to re-expose the json schemas as well?

ramirotw replied:
For this case exposing the compiled schema as in `@cowprotocol/app-data/schemas/v0.4.0.json` should allow the react-jsonschema-form lib to create dynamically the form without defining it manually as @matextrem is doing here.

I tested the the latest compiled schema in react-jsonschema-form's [playground and works great](https://rjsf-team.github.io/react-jsonschema-form/#eyJmb3JtRGF0YSI6eyJtZXRhZGF0YSI6eyJyZWZlcnJlciI6e30sInF1b3RlIjp7fX0sImZpcnN0TmFtZSI6IkNodWNrIiwibGFzdE5hbWUiOiJOb3JyaXMiLCJhZ2UiOjc1LCJiaW8iOiJSb3VuZGhvdXNlIGtpY2tpbmcgYXNzZXMgc2luY2UgMTk0MCIsInBhc3N3b3JkIjoibm9uZWVkIn0sInNjaGVtYSI6eyIkaWQiOiJodHRwczovL2Nvd3N3YXAuZXhjaGFuZ2Uvc2NoZW1hcy9hcHAtZGF0YS92MC40LjAuanNvbiIsIiRzY2hlbWEiOiJodHRwOi8vanNvbi1zY2hlbWEub3JnL2RyYWZ0LTA3L3NjaGVtYSIsImRlc2NyaXB0aW9uIjoiTWV0YWRhdGEgSlNPTiBkb2N1bWVudCBmb3IgYWRkaW5nIGluZm9ybWF0aW9uIHRvIG9yZGVycy4iLCJyZXF1aXJlZCI6WyJ2ZXJzaW9uIiwibWV0YWRhdGEiXSwidGl0bGUiOiJBcHBEYXRhIFJvb3QgU2NoZW1hIiwidHlwZSI6Im9iamVjdCIsInByb3BlcnRpZXMiOnsidmVyc2lvbiI6eyIkaWQiOiIjL2RlZmluaXRpb25zL3ZlcnNpb24iLCJkZXNjcmlwdGlvbiI6IlNlbWFudGljIHZlcnNpb25pbmcgb2YgZG9jdW1lbnQiLCJleGFtcGxlcyI6WyIxLjAuMCIsIjEuMi4zIl0sInRpdGxlIjoiVmVyc2lvbiIsInR5cGUiOiJzdHJpbmcifSwiYXBwQ29kZSI6eyIkaWQiOiIjL3Byb3BlcnRpZXMvYXBwQ29kZSIsImRlc2NyaXB0aW9uIjoiVGhlIGNvZGUgaWRlbnRpZnlpbmcgdGhlIENMSSwgVUksIHNlcnZpY2UgZ2VuZXJhdGluZyB0aGUgb3JkZXIuIiwiZXhhbXBsZXMiOlsiQ293U3dhcCJdLCJ0aXRsZSI6IkFwcCBDb2RlIiwidHlwZSI6InN0cmluZyJ9LCJlbnZpcm9ubWVudCI6eyIkaWQiOiIjL3Byb3BlcnRpZXMvZW52aXJvbm1lbnQiLCJkZXNjcmlwdGlvbiI6IkVudmlyb25tZW50IGZyb20gd2hpY2ggdGhlIG9yZGVyIGNhbWUgZnJvbSIsInRpdGxlIjoiRW52aXJvbm1lbnQiLCJ0eXBlIjoic3RyaW5nIiwiZXhhbXBsZXMiOlsicHJvZHVjdGlvbiIsImRldmVsb3BtZW50Iiwic3RhZ2luZyIsImVucyJdfSwibWV0YWRhdGEiOnsiJGlkIjoiIy9wcm9wZXJ0aWVzL21ldGFkYXRhIiwiZGVmYXVsdCI6e30sImRlc2NyaXB0aW9uIjoiRWFjaCBtZXRhZGF0YSB3aWxsIHNwZWNpZnkgb25lIGFzcGVjdCBvZiB0aGUgb3JkZXIuIiwicmVxdWlyZWQiOltdLCJ0aXRsZSI6Ik1ldGFkYXRhIiwidHlwZSI6Im9iamVjdCIsInByb3BlcnRpZXMiOnsicmVmZXJyZXIiOnsiJGlkIjoiI3JlZmVycmVyL3YwLjEuMC5qc29uIiwiJHNjaGVtYSI6Imh0dHA6Ly9qc29uLXNjaGVtYS5vcmcvZHJhZnQtMDcvc2NoZW1hIiwicmVxdWlyZWQiOlsidmVyc2lvbiIsImFkZHJlc3MiXSwidGl0bGUiOiJSZWZlcnJlciIsInR5cGUiOiJvYmplY3QiLCJwcm9wZXJ0aWVzIjp7InZlcnNpb24iOnsiJHJlZiI6IiMvcHJvcGVydGllcy92ZXJzaW9uIn0sImFkZHJlc3MiOnsidGl0bGUiOiJSZWZlcnJlciBhZGRyZXNzIiwiJGlkIjoiIy9kZWZpbml0aW9ucy9ldGhlcmV1bUFkZHJlc3MiLCJwYXR0ZXJuIjoiXjB4W2EtZkEtRjAtOV17NDB9JCIsImV4YW1wbGVzIjpbIjB4YjZCQWQ0MWFlNzZBMTFEMTBmN2IwRTY2NEM1MDA3YjkwOGJDNzdDOSJdLCJ0eXBlIjoic3RyaW5nIn19fSwicXVvdGUiOnsiJGlkIjoiI3F1b3RlL3YwLjIuMC5qc29uIiwiJHNjaGVtYSI6Imh0dHA6Ly9qc29uLXNjaGVtYS5vcmcvZHJhZnQtMDcvc2NoZW1hIiwicmVxdWlyZWQiOlsidmVyc2lvbiIsInNsaXBwYWdlQmlwcyJdLCJ0aXRsZSI6IlF1b3RlIiwidHlwZSI6Im9iamVjdCIsInByb3BlcnRpZXMiOnsidmVyc2lvbiI6eyIkcmVmIjoiIy9wcm9wZXJ0aWVzL3ZlcnNpb24ifSwic2xpcHBhZ2VCaXBzIjp7IiRpZCI6IiMvcHJvcGVydGllcy9zbGlwcGFnZUJpcHMiLCJ0aXRsZSI6IlNsaXBwYWdlIEJpcHMiLCJkZXNjcmlwdGlvbiI6IlNsaXBwYWdlIHRvbGVyYW5jZSB0aGF0IHdhcyBhcHBsaWVkIHRvIHRoZSBvcmRlciB0byBnZXQgdGhlIGxpbWl0IHByaWNlLiBFeHByZXNzZWQgaW4gQmFzaXMgUG9pbnRzIChCSVBTKSIsImV4YW1wbGVzIjpbIjUiLCIxMCIsIjIwIiwiMTAwIl0sInBhdHRlcm4iOiJeXFxkKyhcXC5cXGQrKT8kIiwidHlwZSI6InN0cmluZyJ9fX19fX19LCJ1aVNjaGVtYSI6eyJmaXJzdE5hbWUiOnsidWk6YXV0b2ZvY3VzIjp0cnVlLCJ1aTplbXB0eVZhbHVlIjoiIiwidWk6YXV0b2NvbXBsZXRlIjoiZmFtaWx5LW5hbWUifSwibGFzdE5hbWUiOnsidWk6ZW1wdHlWYWx1ZSI6IiIsInVpOmF1dG9jb21wbGV0ZSI6ImdpdmVuLW5hbWUifSwiYWdlIjp7InVpOndpZGdldCI6InVwZG93biIsInVpOnRpdGxlIjoiQWdlIG9mIHBlcnNvbiIsInVpOmRlc2NyaXB0aW9uIjoiKGVhcnRoaWFuIHllYXIpIn0sImJpbyI6eyJ1aTp3aWRnZXQiOiJ0ZXh0YXJlYSJ9LCJwYXNzd29yZCI6eyJ1aTp3aWRnZXQiOiJwYXNzd29yZCIsInVpOmhlbHAiOiJIaW50OiBNYWtlIGl0IHN0cm9uZyEifSwiZGF0ZSI6eyJ1aTp3aWRnZXQiOiJhbHQtZGF0ZXRpbWUifSwidGVsZXBob25lIjp7InVpOm9wdGlvbnMiOnsiaW5wdXRUeXBlIjoidGVsIn19fSwidGhlbWUiOiJkZWZhdWx0IiwibGl2ZVNldHRpbmdzIjp7InZhbGlkYXRlIjpmYWxzZSwiZGlzYWJsZSI6ZmFsc2UsInJlYWRvbmx5IjpmYWxzZSwib21pdEV4dHJhRGF0YSI6ZmFsc2UsImxpdmVPbWl0IjpmYWxzZX19)

We should be exposing an utility method in the [Metadata API](https://github.com/cowprotocol/cow-sdk/blob/main/src/api/metadata/index.ts) like `getSchema(version: string = LATEST_APP_DATA_VERSION)`

ramirotw replied:
btw, was revisiting the main CowSdk file and we still have there  the [`validateAppDataDocument`](https://github.com/cowprotocol/cow-sdk/blob/main/src/CowSdk.ts#L42) method. Shouldn't we move that to the [Metadata API](https://github.com/cowprotocol/cow-sdk/blob/main/src/api/metadata/index.ts)?

alfetopito replied:
Hmm, true. There is definitely more stuff to be hoisted up from the SDK to app-data package and exposed via API methods.
Let's discuss more in detail during the sync meeting

matextrem replied:
@alfetopito Is it possible to change schema to looks like the one I have in` config.tsx` ? Since we need to add some dependencies and some other keys in order for us to allow enable/disable metadata fields

ramirotw replied:
what if if we just decorate the json schema with the props that we need? I don't think updating the original schema with UI stuff is a good approach as it will impact all consumers

alfetopito replied:
I second Ramiro. Having anything else in the schema will signal that it could be included in the appData, which is not what we want.
If extra data is needed it can be iterated over and enhanced here, but it does not belong in the exported schema definitions

alfetopito replied:
New methods on SDK https://github.com/cowprotocol/cow-sdk/pull/54

ramirotw commented on tsconfig.json:
```diff
@@ -29,12 +23,7 @@
-      "webpack-env",
-      "resize-observer-browser"
-    ],
+    "types": ["jest", "node", "webpack-env", "resize-observer-browser"],
```
prettier made these changes?

matextrem replied:
Yes :)

elena-zh commented:
Hey, this is my early review of the PR. 
I know, that UI has not been implemented yet, but anyways I will leave here some my proposals. Maybe they might be useful: 
1. I think it would be nice to add hints as tooltips near a field's label
![image](https://user-images.githubusercontent.com/70885163/180213062-43df77d9-86c9-48ba-ae55-67cb0c26dc12.png)
2. As for environments hint, we could add 'pr' to the list, as it is also used in Cowswap app for appdata
![pr](https://user-images.githubusercontent.com/70885163/180213283-9debf1aa-3015-4514-b059-abeead0e186a.jpg)
3. I think it would be great to show errors in red. Besides, I think it is better to validate fields on FE side first. WDYT?
![image](https://user-images.githubusercontent.com/70885163/180213500-d4387607-192b-48db-8565-01f5f5f92f9e.png)
4. Nice to adjust the page to look better in a mobile view

And here is the list of issues that can be fixed before the UI part: 

1. Based on the issue description, a user can check the Quote or Refferrer section: `Then, for the metadata, user can pick any of the choices of referrer and quote. Maybe a checkbox to enable them?` Currently, both sections are enabled.
2. Quote version field name is incorrect (shows 'quote' instead of the 'version'. The field should be mandatory and show the latest version (read-only)
![field name](https://user-images.githubusercontent.com/70885163/180214001-7cd2fa67-beb2-4718-ab27-d96d44a6d0b6.jpg)
3. Address field should be marked as mandatory field when Referrer section is selected
![image](https://user-images.githubusercontent.com/70885163/180214520-d8e4c1c1-a653-486c-a978-b29f9530bfea.png)
4. There is no validation on the address field when I use upper or lower-case letters. I'm not sure, however, if it is needed. But I have to notice that this validation exists in Explorer when search for a user details
5. Again, I'm not sure if it is an issue, but the fields have a different order in 'referrer' section in a metadata info
![change places](https://user-images.githubusercontent.com/70885163/180215550-31d0a92f-15b2-47a2-96cb-2aaec947ec96.jpg)
6. Slippage Bips field should be marked as a mandatory one (when the section is selected)
7. I think we should make a user aware of errors when there is something wrong with uploading data to IPFS (show an error banner somehow). 
![error somehow](https://user-images.githubusercontent.com/70885163/180216006-7d9caa91-6fcc-49a8-8b48-d8bd16fb4617.jpg)

9. The same with a success notification (show 'successfully uploaded' message/banner or so)

Thanks

ramirotw commented:
It would be helpful to add a description field on the form using the description prop from the schema:

<img width="805" alt="image" src="https://user-images.githubusercontent.com/3328006/180339426-49aef53c-d800-42cc-b5f6-27903ed9e7df.png">

ramirotw commented on src/apps/explorer/pages/Metadata/config.ts:
```diff
@@ -0,0 +1,144 @@
+  if (enableReferrer && !isAddress(address?.toLowerCase())) {
+    errors.metadata['referrer'].address.addError('This is not an address')
+  }
+  return errors
```
why not leverage the `pattern` field the current json schema has? The lib uses ajv for validation so it works the same as the validation we do in the sdk:

```json
"address": {
  "title": "Referrer address",
  "$id": "#/definitions/ethereumAddress",
  "pattern": "^0x[a-fA-F0-9]{40}$",
  "examples": [
    "0xb6BAd41ae76A11D10f7b0E664C5007b908bC77C9"
  ],
  "type": "string"
}
```

docs: https://react-jsonschema-form.readthedocs.io/en/latest/usage/validation/

alongoni commented on src/apps/explorer/pages/Metadata/styled.ts:
```diff
@@ -0,0 +1,120 @@
+      appearance: none;
+      align-items: center;
+      color: ${({ theme }): string => theme.greyShade};
+      padding: 1.6rem 1.6rem 1.6rem 5rem;
```
we can reduce the left padding here to:
`padding: 1.6rem;`

![image](https://user-images.githubusercontent.com/622217/180459485-dc745281-21c8-40bd-9474-fa496ec2b16c.png)

matextrem commented on src/apps/explorer/pages/Metadata/config.ts:
```diff
@@ -0,0 +1,144 @@
+
+export const uiSchema: UiSchema = {
+  environment: {
+    'ui:help': 'Hint: Development, Production, Staging.',
```
It needs to be updated in the appData schema module first in order for this to be a dropdown.

elena-zh commented:
Hey @matextrem , the page looks much-much nicer now!

Just some notes/questions/issues from my side: 
1. I'm not sure about it, but maybe we should not show 'referrer' and 'quote' blocks in the metadata json when nothing is checked in the metadata form? See: the image contains metadata for one of orders, and there is no 'referrer' and 'quote'  info in it
![do not show when not checked](https://user-images.githubusercontent.com/70885163/180792263-aee49a1f-2192-4521-871f-3f29fbf498db.jpg)
2. Not sure at all, but I'd better write this point: 'generate' button is enabled when I open the page, and I do not fill in any fields. Is this OK? Maybe we should disabled it at all, otherwise the metadata doc will contain version only. @alfetopito , WDYT?
![уьзен](https://user-images.githubusercontent.com/70885163/180793714-599b557a-7d70-4005-aa05-ac199be86e85.jpg)

3. I still vote for showing validation for fields in the FE. It will result in the same style of error messages: now we have 2 types that come from BE
![image](https://user-images.githubusercontent.com/70885163/180793014-b02d7029-520f-427f-8f33-d823dc3a9c19.png)
![image](https://user-images.githubusercontent.com/70885163/180793209-804ae2c0-8d17-4f56-824b-161a16fc2015.png)

4. Continuation of the 3rd point: adding a validation of FE will help to keep 'Generate' button disabled when this or that field is filled in with invalid data. 
5. Also it will help to remove a previous error message when clear out the field and then will in with a valid value
![image](https://user-images.githubusercontent.com/70885163/180793938-fc13f9ff-bab0-47f3-962c-8db942e94fb0.png)

6. If leave BE errors, then it would be great to make them a bit more readable (show something like 'Only numbers accepted' or so instead of should match pattern "^\d+(\.\d+)?$"), and start a message from the capital letter
![er expl](https://user-images.githubusercontent.com/70885163/180794995-000fafff-4c57-465f-8f6e-567ce9695bd9.jpg)

7. Error messages do not appear when: open the form, press on the 'Generate' button, then change field's values in the form and press on the "Generate' button once again
![no errors](https://user-images.githubusercontent.com/70885163/180795407-8cecf5c3-e050-4aed-b396-ab7f0929afad.jpg)

8. I would be great to clear out the json doc when uncheck 'referrer' and 'quote' sections inside the metadata doc
![uncheck](https://user-images.githubusercontent.com/70885163/180795603-d8feb889-3ea6-43de-b94d-d96925dbde82.jpg)

9. When I'm filling in the Slippage Bips* field in the 'Quote' section, it is moving
10. Still it would be nice to make a mobile view look better: 
![image](https://user-images.githubusercontent.com/70885163/180795890-d3bfbf75-7ad0-42df-95dc-af6766523c62.png)
![mobile](https://user-images.githubusercontent.com/70885163/180795995-df16a8a6-b596-471b-9ba0-ef08478f7408.jpg)

https://user-images.githubusercontent.com/70885163/180796048-a161a88e-2cd2-4c67-bbeb-4d4723a868d0.mp4

11. How can I upload the file (where I should fill in IPFS credentials, and where can I get them?)
![how](https://user-images.githubusercontent.com/70885163/180796193-269abef9-b492-41b6-931c-4659b814ad0a.jpg)


Thanks!

matextrem commented:
> Hey @matextrem , the page looks much-much nicer now!
> 
> Just some notes/questions/issues from my side:
> 
> 1. I'm not sure about it, but maybe we should not show 'referrer' and 'quote' blocks in the metadata json when nothing is checked in the metadata form? See: the image contains metadata for one of orders, and there is no 'referrer' and 'quote'  info in it
>    ![do not show when not checked](https://user-images.githubusercontent.com/70885163/180792263-aee49a1f-2192-4521-871f-3f29fbf498db.jpg)
> 2. Not sure at all, but I'd better write this point: 'generate' button is enabled when I open the page, and I do not fill in any fields. Is this OK? Maybe we should disabled it at all, otherwise the metadata doc will contain version only. @alfetopito , WDYT?
>    ![уьзен](https://user-images.githubusercontent.com/70885163/180793714-599b557a-7d70-4005-aa05-ac199be86e85.jpg)
> 3. I still vote for showing validation for fields in the FE. It will result in the same style of error messages: now we have 2 types that come from BE
>    ![image](https://user-images.githubusercontent.com/70885163/180793014-b02d7029-520f-427f-8f33-d823dc3a9c19.png)
>    ![image](https://user-images.githubusercontent.com/70885163/180793209-804ae2c0-8d17-4f56-824b-161a16fc2015.png)
> 4. Continuation of the 3rd point: adding a validation of FE will help to keep 'Generate' button disabled when this or that field is filled in with invalid data.
> 5. Also it will help to remove a previous error message when clear out the field and then will in with a valid value
>    ![image](https://user-images.githubusercontent.com/70885163/180793938-fc13f9ff-bab0-47f3-962c-8db942e94fb0.png)
> 6. If leave BE errors, then it would be great to make them a bit more readable (show something like 'Only numbers accepted' or so instead of should match pattern "^\d+(.\d+)?$"), and start a message from the capital letter
>    ![er expl](https://user-images.githubusercontent.com/70885163/180794995-000fafff-4c57-465f-8f6e-567ce9695bd9.jpg)
> 7. Error messages do not appear when: open the form, press on the 'Generate' button, then change field's values in the form and press on the "Generate' button once again
>    ![no errors](https://user-images.githubusercontent.com/70885163/180795407-8cecf5c3-e050-4aed-b396-ab7f0929afad.jpg)
> 8. I would be great to clear out the json doc when uncheck 'referrer' and 'quote' sections inside the metadata doc
>    ![uncheck](https://user-images.githubusercontent.com/70885163/180795603-d8feb889-3ea6-43de-b94d-d96925dbde82.jpg)
> 9. When I'm filling in the Slippage Bips* field in the 'Quote' section, it is moving
> 10. Still it would be nice to make a mobile view look better:
>     ![image](https://user-images.githubusercontent.com/70885163/180795890-d3bfbf75-7ad0-42df-95dc-af6766523c62.png)
>     ![mobile](https://user-images.githubusercontent.com/70885163/180795995-df16a8a6-b596-471b-9ba0-ef08478f7408.jpg)
> 
>  moving.mp4 
> 11. How can I upload the file (where I should fill in IPFS credentials, and where can I get them?)
>     ![how](https://user-images.githubusercontent.com/70885163/180796193-269abef9-b492-41b6-931c-4659b814ad0a.jpg)
> 
> Thanks!

@elena-zh 
1. We show them since it is returned by the schema and given the fact it is still a valid metadata document we add them. We could just add a logic in order to remove them but IDK if this is really worth it. WDYT @alfetopito 

2. I think and empty metadata document with only the version is still a valid one. 

3.  Validations are happening on the FE side. We are not calling any backend at all. 

4. I think that would decrease the performance by checking for every field on data change. CC: @ramirotw @alfetopito 

5. Fixed

6. Fixed

7. Fixed

8. Fixed

9 and 10. Maybe @alongoni could give us a hand here. 

11. I think we should add it in the REPO settings, right? @alfetopito

elena-zh commented:
Hey @matextrem , thank you!
So some notes: case 3. I assumed that errors are from BE side as they appear only after when I press on the 'Generate' button (and now it works this way). So here is my point is to show errors when a field is filled in with an invalid value without need to press on the button

Then: 
1. I can see a 'double' error for the referrer field
![double](https://user-images.githubusercontent.com/70885163/180981084-04af465e-b814-4081-92fe-b759945180cb.jpg)
2. I would rewrite 'Required field' error message, and start it with the capital letter
![image](https://user-images.githubusercontent.com/70885163/180981219-86cdad99-cbc6-4a5c-baed-f707cea555c5.png)
3. I would apply the same validation message to the Slippage field (not it does not show message that the field is required)
![image](https://user-images.githubusercontent.com/70885163/180981387-51b8f74f-a9c6-4902-b30c-b31756372ded.png)

Then, let wait for @alongoni fixes for UI part, and answers from @alfetopito and @ramirotw for the points mentioned above. 

Thanks!

matextrem commented:
@elena-zh Comments fixed!

alfetopito commented:
Will start by addressing Elena's/Mati's comments https://github.com/cowprotocol/explorer/pull/168#issuecomment-1194677566



> 1. We show them since it is returned by the schema and given the fact it is still a valid metadata document we add them. We could just add a logic in order to remove them but IDK if this is really worth it. WDYT @alfetopito

Agree with Elena, even though it might be valid I rather not have empty metadatas (apart of course from the required `metadata: {}` field)

> 2. I think and empty metadata document with only the version is still a valid one.

Indeed, only version and metadata is the minimal valid document!
Still, I would show the button only after the form becomes valid, if possible.
Meaning, only after the form is touched AND the result is valid.

> 3. Validations are happening on the FE side. We are not calling any backend at all.

The confusion here is that there are 2 validations being applied. The first is from the browser for default HTML validation and the other for logic validation.
To keep them consistent then we should disable HTML validation and make sure those are covered in the logic validation.

> 4. I think that would decrease the performance by checking for every field on data change. CC: @ramirotw @alfetopito

I would think this is a pretty cheap operation performance wise. Not the case?
One alternative is to debounce the validation. 
If this is still too heavy then I agree on keeping it as Mati proposes

> 11. I think we should add it in the REPO settings, right? @alfetopito

We discussed this awhile ago and the idea was to allow people to upload it here, given they provide their Pinata credentials.
I would say that is a follow up PR, and for now this option can be hidden/removed

alfetopito commented:
It looks and works great! Really like how this is shaping up, awesome job 👏 


Some notes about the page/behaviour

1. We should make the width fixed 

https://user-images.githubusercontent.com/43217/181231809-7789b096-dfa6-4914-ae6e-ec87e7d33a03.mov

2. As a user, I don't know much what is going on. We need to add context in the form of titles, sections, tooltips, links to FAQ/docs/etc. Like, what do I do with this? If I copy this JSON, where should I put it? Why do I want to upload this to IPFS?

Will not review the code

alfetopito commented on src/apps/explorer/ExplorerApp.tsx:
```diff
@@ -132,6 +140,7 @@ const AppContent = (): JSX.Element => {
           <Route path={pathPrefix + '/address/:address'} exact component={UserDetails} />
           <Route path={pathPrefix + '/tx/:txHash'} exact component={TransactionDetails} />
           <Route path={pathPrefix + '/search/:searchString?'} exact component={SearchNotFound} />
+          <Route path={pathPrefix + '/metadata'} exact component={MetadataDetails} />
```
cc @alongoni to add to the redesigned menu you are working on

alfetopito commented on src/apps/explorer/pages/Metadata/config.ts:
```diff
@@ -0,0 +1,214 @@
+export const INITIAL_FORM_VALUES = {
+  version: LATEST_APP_DATA_VERSION,
+  metadata: {
+    referrer: {},
+    quote: {},
```
As mentioned in the comments, they should not be present if not used

alfetopito commented on src/apps/explorer/pages/Metadata/config.ts:
```diff
@@ -0,0 +1,214 @@
+}
+
+export const getSchema = async (): Promise<JSONSchema7> => {
+  const { default: latestSchema } = await import(`@cowprotocol/app-data/schemas/v${LATEST_APP_DATA_VERSION}.json`)
```
There's an util for this, exposed on app-data and sdk 
https://github.com/cowprotocol/app-data/blob/main/src/utils.ts#L42=
https://github.com/cowprotocol/cow-sdk/blob/main/src/api/metadata/index.ts#L53=

alfetopito commented on src/apps/explorer/pages/Metadata/config.ts:
```diff
@@ -0,0 +1,214 @@
+const formatSchema = (schema: JSONSchema7): JSONSchema7 => {
+  const formattedSchema = structuredClone(schema)
+
+  deletePropertyPath(formattedSchema, 'properties.appCode.examples')
+  deletePropertyPath(formattedSchema, 'properties.version.examples')
+  deletePropertyPath(formattedSchema, 'properties.environment.examples')
+  deletePropertyPath(formattedSchema, 'properties.metadata.properties.referrer.properties.address.examples')
+  deletePropertyPath(formattedSchema, 'properties.metadata.properties.quote.properties.slippageBips.examples')
```
Since you are already navigating through the schema, what about writing a method to remove a property by name:

```js
deleteAllPropertiesByName(schema, 'examples')
```

alfetopito reviewed (commented):
Some minor comments on the code.

Overall, it is way too coupled with the schema definition and will require maintenance for every schema change.
I understand this is part of the requirements we gave you, but I'm not happy with the direction.
The idea was to make it so the schema would make our life easier, not harder.

We need to discuss if there are any alternatives to decouple without requiring so much fine tuning.

elena-zh commented:
Hey @matextrem , the address field LGTM now!
However, the slippage field validation message is broken again: 
![image](https://user-images.githubusercontent.com/70885163/181267065-c06147d1-e081-49af-9417-7944db8a51d3.png)

Then, I noticed that Slippage field is expanded with existing validation message when I fill in 'Referrer' address once. And vice versa: if I fill in slippage, then expand a Referrer section, then I will see 'Address' field with validation message (however, I have not touched it). See the [video](https://watch.screencastify.com/v/171FiDlNbmtBqz4s8Lg7). Could we please fix it somehow?

Thanks!

matextrem commented:
@elena-zh 
I've fixed the slippage field validation message point.
Related to the second one you mentioned, it's not possible to do that since validation depends on schema and given the fact we have the referrer address and Slippage bips as required, when you select any of them, validations rules are applied. (The fact that you don't see it the first time it is because of the validation being triggered when submit button is pressed) - You can see an example [here](https://rjsf-team.github.io/react-jsonschema-form/#eyJmb3JtRGF0YSI6eyJmaXJzdE5hbWUiOmZhbHNlfSwic2NoZW1hIjp7InR5cGUiOiJvYmplY3QiLCJwcm9wZXJ0aWVzIjp7ImZpcnN0TmFtZSI6eyJ0eXBlIjoiYm9vbGVhbiJ9fSwiZGVwZW5kZW5jaWVzIjp7ImZpcnN0TmFtZSI6eyJvbmVPZiI6W3sicHJvcGVydGllcyI6eyJmaXJzdE5hbWUiOnsiZW51bSI6W3RydWVdfSwibGFzdE5hbWUiOnsidHlwZSI6InN0cmluZyJ9fX1dfX19LCJ1aVNjaGVtYSI6e30sInRoZW1lIjoiYm9vdHN0cmFwLTQiLCJsaXZlU2V0dGluZ3MiOnsidmFsaWRhdGUiOmZhbHNlLCJkaXNhYmxlIjpmYWxzZSwib21pdEV4dHJhRGF0YSI6ZmFsc2UsImxpdmVPbWl0IjpmYWxzZX19) of how it works

However, I tried to do some hacky solution but it didn't work since it breaks the validation functionality.

elena-zh commented:
Hey @matextrem , great, thank you!
Just some notes based on @alfetopito 's comment above: 

> Indeed, only version and metadata is the minimal valid document!
> Still, I would show the button only after the form becomes valid, if possible.
> Meaning, only after the form is touched AND the result is valid.

1. Since the doc is valid with 1 version field, I think it would be nice to enable 'Generate' button when touch any optional field (see the [video ](https://watch.screencastify.com/v/9nJhH6NjNOu55WQMB7Tc) what I mean). @alfetopito , am I correct, or implemented behavior looks good to you?
2. Then, I see that 'Generate' button is enabled when address/spillage fields are filled in with an invalid value/empty
![image](https://user-images.githubusercontent.com/70885163/181487120-679e60d8-cbbc-4156-aa3f-f8487083fe2d.png)
3. Just to be on the same page: we leave the 1st validation as it is (On 'Generate' button click), correct? @matextrem , @ramirotw 

> As a user, I don't know much what is going on. We need to add context in the form of titles, sections, tooltips, links to FAQ/docs/etc. Like, what do I do with this? If I copy this JSON, where should I put it? Why do I want to upload this to IPFS?

@alfetopito , should I create a separate task for this? And who can help to generate content for all these things?

> We discussed this awhile ago and the idea was to allow people to upload it here, given they provide their Pinata credentials.
> I would say that is a follow up PR, and for now this option can be hidden/removed

@matextrem added these fields to the form 
![image](https://user-images.githubusercontent.com/70885163/181487816-4cf4b78f-75bb-4280-bedc-603bc4dbc5a1.png)

but still I do not have credentials to upload data

As for the UI issues, still need to wait for @alongoni changes in the PR. 
However, I would add this one: I think that the 'Upload' button should be the same as the 'Generate' one
![image](https://user-images.githubusercontent.com/70885163/181488105-3159c3fc-f358-40e9-9fae-e9b7d953a384.png)

Also, I think we need to add a min validation rules to these fields, and the button (keep it disabled until fill in a field/fields with a valid data)
![image](https://user-images.githubusercontent.com/70885163/181488336-6ba1e43b-83fa-438f-9d7e-0bf320ca3466.png)

Thanks

elena-zh commented:
Hey @matextrem , @alongoni , cool changes!!

Some notes from my side for @alongoni : 
1. json section is not fixed on the formed and is [moving ](https://watch.screencastify.com/v/Vtf9xsKWIgaymg6FCsAJ)when scroll the page. Could you please fix it at the top, + display it in 1 line with the main form? 
![fix in 1 line](https://user-images.githubusercontent.com/70885163/182151436-69e6a3ab-1ea3-4715-aa78-f88a6b40b456.jpg)

2. Generate button becomes too wide when press on it
![generate button expanded](https://user-images.githubusercontent.com/70885163/182151707-909a43de-86de-4825-8670-1f9769856e24.jpg)

3. We have some issue with overlapping elements on the page when screen width is [736 ~ 1020] px
![tablet](https://user-images.githubusercontent.com/70885163/182151904-d03686a0-9c1c-4ee3-9e86-4f36297e7821.jpg)
![tablet1](https://user-images.githubusercontent.com/70885163/182151906-485b48a2-702c-4bcd-baf8-01fe61001f1d.jpg)
![tablet2](https://user-images.githubusercontent.com/70885163/182151909-52250b1a-fcbe-4c74-82b6-e2b5cf3e100a.jpg)
![tablet3](https://user-images.githubusercontent.com/70885163/182151911-9aa83cb7-4833-4674-81a8-11237d193fe4.jpg)

4. Error banner is too big in the desktop version, and is clued to the button in a mobile view:
![banner](https://user-images.githubusercontent.com/70885163/182152110-113f7e2f-5913-4df4-9777-253b433078c0.jpg)
![mobile banner](https://user-images.githubusercontent.com/70885163/182152112-5d7101cc-dbc2-4177-aeba-e49c4eee9c80.jpg)

5. Root schema boarder does not reaches the section name in iOS mobile device
![image](https://user-images.githubusercontent.com/70885163/182152652-ee2459cd-12d0-4f83-b644-293d29290ff0.png)

6. Sections Pinata section header is displayed inside the section (not on the boarder)
![pinata diff design](https://user-images.githubusercontent.com/70885163/182152931-e2182ea3-eb6f-4bcc-87e6-9a4f2b67b496.jpg)
7. It would be great whether to add or remove periods in all sections' descriptions: currently some of them have a period, some not
![period](https://user-images.githubusercontent.com/70885163/182153124-5e6d7a67-99f7-4a31-8b3a-3a906da98909.jpg)
8. Maybe a question to @alfetopito as well, but it would be nice to update tooltips for these fields: 
![tooltip](https://user-images.githubusercontent.com/70885163/182153234-243fd93d-8d81-4b98-9f7e-708b977a7c88.jpg)
![tooltips 2](https://user-images.githubusercontent.com/70885163/182153235-acf24a74-1671-4f84-8f88-08c382fa38fe.jpg)


Some issues for @matextrem : 
1. It would be nice to remove empty fields from json when clear out appcode and environment fields
![remove when claer out](https://user-images.githubusercontent.com/70885163/182153444-725485d6-bf30-49d7-8592-48b447634562.jpg)
2. When I clear out slippage and address fields, I see other validation messages instead of 'Required field'
![empty](https://user-images.githubusercontent.com/70885163/182153688-0d9803c5-a61a-4223-aecb-033ad0c33aeb.jpg)
3. I do not see validation message + disabled 'Generate' button when I change an address to an invalid one after the appdata doc has been generated (see the [video](https://watch.screencastify.com/v/s8L4msclTY7wkJK6EFx5)). Maybe we should keep 'Generate' button disabled after the doc is generated until we change a form with other valid data, WDYT?
Thanks!

alongoni commented:
thanks @elena-zh for the feedback!
I think having the json section floating makes it easier for users to follow the updates as they fill out the form. WDYT @alfetopito ?

Regarding point 7 (missing periods) I'll be adding them here: https://github.com/cowprotocol/app-data/blob/main/src/schemas/v0.4.0.json

elena-zh commented:
Hey @matextrem and @alongoni !
Great job! The form looks nicer and nicer!
Just to recap what is left:
1. Tooltips
2. AppData Root Schema boarder in iOS (but I can create a separate issue for this)
3. Periods in the end of sentences (https://github.com/cowprotocol/app-data/blob/main/src/schemas/v0.4.0.json)

New issues I have found: 
1. I see 'Type error' message in iOS when I fill in Pinata fields with an invalid data (I expect to see 'Invalid API key provided' message)
![image](https://user-images.githubusercontent.com/70885163/182410211-75ddf092-9ff6-4b08-bafc-aa2d8934797a.png)

2.  Error messages in iOS have 'additional pixels' . See the image
![image](https://user-images.githubusercontent.com/70885163/182410464-348f9399-5a87-4f36-ae3f-e697121e02d9.png)
![image](https://user-images.githubusercontent.com/70885163/182410699-35e6aab4-2089-48f3-a2c8-50dda4b61b9b.png)

3. When I press on the 'Upload button', an error log appears in the console
![error in console](https://user-images.githubusercontent.com/70885163/182410804-b50ca6dd-01b1-42f8-a4d6-302765f2604d.jpg)

4. Upload button is enabled once a user enters Pinata credentials. So the issue is that when a user changes the Appdata form, skips pressing on the 'Generate' button and hits 'Upload', the previous version of the doc is uploaded, not the updated one. 
Can we do anything with this? I mean, disable 'Upload' button unless a user presses 'Generate' button once again or so? 
![image](https://user-images.githubusercontent.com/70885163/182411462-c4d4ecff-1943-489c-9041-b63d6a563e48.png)

Thanks!

elena-zh commented:
Hey @alongoni , fields look great now!

Hey @matextrem , love these changes!
I have faced some 'side issues', I think: 
1. When I open an uploaded doc, I see a lot of useless info in it
![schema](https://user-images.githubusercontent.com/70885163/182574010-20792940-2f8e-4a8d-9f0d-a2a3141984e5.jpg)
2. When I start editing the form after upload, I still see copy icon and successful banner
![hide](https://user-images.githubusercontent.com/70885163/182574279-a5b23cab-f886-466d-9ffd-e9c2f8a7894e.jpg)
3. When I regenerate a doc and upload it, I can't open this doc =(
![image](https://user-images.githubusercontent.com/70885163/182574417-cb0c1bbf-090a-45f7-8af7-7fc7b5060137.png)

Could you please take a look into these issues?

Thanks

elena-zh commented:
Great changes!
I have reported #174 issue for UI issues in iOS related to the current page. 

Then, the issue with periods is not yet implemented. 

**And very important case**: we still do not have a navigation link to open the page on Explorer. 

Some issues with a new 'Decode' section: 
1. It would be great to make it look better in a mobile view: 
![image](https://user-images.githubusercontent.com/70885163/182868843-fc8ae9f2-0176-4271-9c4a-743c6608d500.png)
![image](https://user-images.githubusercontent.com/70885163/182869971-5e759e3a-0546-4c1b-a83c-09a7e49cbdf7.png)


2. Maybe we should add some validation rules to the App Data hash field? 
![image](https://user-images.githubusercontent.com/70885163/182869271-8f949dcf-34c3-4be2-b514-b067d330276d.png)

3. When I press on the Decode button, the app data hash appears as a link witn a show more button. Maybe we should show it already expanded as it repeats the hash we already pasted into the field?
![expnadned](https://user-images.githubusercontent.com/70885163/182870029-c6627190-503a-4509-91b8-56b223141102.jpg)
4. When the data is expanded, I'd make the json section left-aligned
![allign](https://user-images.githubusercontent.com/70885163/182870219-e423ee00-daf4-4939-978a-1f4240a92334.jpg)

Thanks!

matextrem commented:
@elena-zh Fixed!

Regarding point number 2, we have required validation for appData hash field. IDK what others validations could add. Thoughts, @alfetopito ?

elena-zh commented:
> @elena-zh Fixed!
> 
> Regarding point number 2, we have required validation for appData hash field. IDK what others validations could add. Thoughts, @alfetopito ?

Hey @matextrem , maybe length at least?
![image](https://user-images.githubusercontent.com/70885163/183027521-bbd49a88-2a04-4ca7-bfc1-abe82b830f41.png)

elena-zh commented:
Some notes: 
1. I still would make all elements left-aligned for the decoded section in the Decode part
![left](https://user-images.githubusercontent.com/70885163/183038200-2114d56b-32f9-454c-b6c8-3a38886671fc.jpg)
![left1](https://user-images.githubusercontent.com/70885163/183038201-4848fc47-1821-4a16-b9ec-09965d3030ef.jpg)
![left-aligned](https://user-images.githubusercontent.com/70885163/183038205-7b462196-b463-44cc-b3cc-036935960887.jpg)
2. Json section in the Form and Decode sections still does not look good in a mobile view
![image](https://user-images.githubusercontent.com/70885163/183039295-9f5e08cb-a421-4778-ba96-91e421d64660.png)
![image](https://user-images.githubusercontent.com/70885163/183039359-a0fa9aa5-3af5-44b1-889b-c0587cafc95a.png)

3. And just an enhancement, I'd like to propose not to reset values in the fields when switch between tabs. @alfetopito , @matextrem , @ramirotw WDYT? It is not a task for this PR, so if we agree to implement it, I will open a separate issue for this. 

Thanks

alfetopito commented:
> @elena-zh Fixed!
> 
> Regarding point number 2, we have required validation for appData hash field. IDK what others validations could add. Thoughts, @alfetopito ?

The hash regex should suffice. Out of my head, something like: `^0x[a-fA-F0-9]{64}`

If it doesn't match that, we can't decode it.

alfetopito commented:
> 3. And just an enhancement, I'd like to propose not to reset values in the fields when switch between tabs. @alfetopito , @matextrem , @ramirotw WDYT? It is not a task for this PR, so if we agree to implement it, I will open a separate issue for this.

Not a big deal to me, but could be addressed as a follow up task.

elena-zh commented:
Hey @matextrem ,  @alongoni changes LGTM!

Still, I see that the issue with periods in the end of sentences is not fixed (will be addressed in https://github.com/cowprotocol/app-data/pull/10 a bit later).
![image](https://user-images.githubusercontent.com/70885163/183433403-e9bc6d32-fdfa-4b2c-aaeb-7557d52f80a7.png)

![image](https://user-images.githubusercontent.com/70885163/183433454-5ff77c0d-0543-4bfb-aa87-e4fbf5dbfc8a.png)

And there is no path to the page, but will be implemented soon. 

Also I have found #175 issue that @ramirotw can take a look into a bit later. 

So I'm approving the current PR.

alfetopito commented on src/apps/explorer/ExplorerApp.tsx:
```diff
@@ -132,6 +140,7 @@ const AppContent = (): JSX.Element => {
           <Route path={pathPrefix + '/address/:address'} exact component={UserDetails} />
           <Route path={pathPrefix + '/tx/:txHash'} exact component={TransactionDetails} />
           <Route path={pathPrefix + '/search/:searchString?'} exact component={SearchNotFound} />
+          <Route path={pathPrefix + '/metadata'} exact component={AppDataDetails} />
```
```suggestion
          <Route path={pathPrefix + '/appdata'} exact component={AppDataDetails} />
```

alfetopito reviewed (commented):
Overall I'm happy with the behaviour and looks good for merging.

Doesn't mean there isn't more stuff to address :)
I'll write it down in more details tomorrow, when I'll also finish the code review.

alfetopito commented on src/apps/explorer/pages/AppData/FormPage.tsx:
```diff
@@ -0,0 +1,267 @@
+    return formattedData
+  }
+
+  const handleMetadataErrors = (_: FormProps, errors: FormValidation): FormValidation =>
+    handleErrors(formRef, errors, setDisabledAppData)
+
+  const handleIPFSErrors = (_: FormProps, errors: FormValidation): FormValidation =>
+    handleErrors(ipfsFormRef, errors, setDisabledIPFS)
```
`useCallback` on the handlers? Most of the inner functions in this component actually

alfetopito commented on src/apps/explorer/pages/AppData/FormPage.tsx:
```diff
@@ -0,0 +1,267 @@
+    if (!network) return
+    setIsLoading(true)
+    try {
+      const hashInfo = await COW_SDK[network]?.metadataApi.calculateAppDataHash(handleFormatData(formData))
```
The appData calculation actually doesn't depend on the network.
You should be safe using Mainnet until your SDK changes are applied to the Explorer

alfetopito commented on src/apps/explorer/pages/AppData/FormPage.tsx:
```diff
@@ -0,0 +1,267 @@
+  }
+
+  const onUploadToIPFS = async ({ formData }: FormProps): Promise<void> => {
+    if (!network || !ipfsHashInfo) return
```
Same here, no need for network

alfetopito commented on src/apps/explorer/pages/AppData/config.tsx:
```diff
@@ -0,0 +1,290 @@
+  const formattedSchema = structuredClone(schema)
+
+  if (formattedSchema?.properties?.version) {
+    formattedSchema.properties.version['readOnly'] = true
+    formattedSchema.properties.version['default'] = LATEST_APP_DATA_VERSION
```
Damn, this looks like the thing that SHOULD be in the schema definition.

If this is what you were pointing out all along Mati, I'm very sorry that I never understood your proposal.

I've created this issue to address that https://github.com/cowprotocol/app-data/issues/11

alfetopito commented on src/apps/explorer/pages/AppData/config.tsx:
```diff
@@ -0,0 +1,290 @@
+      required: requiredFields,
+    }
+    formattedSchema.properties.metadata['properties'].referrer.dependencies = referrerDependencies
+  }
```
Lines 46 to 88 are pretty similar.
Can you extract a function to deal with them?

alfetopito commented on src/apps/explorer/pages/AppData/index.tsx:
```diff
@@ -0,0 +1,71 @@
+  return [
+    {
+      id: TabView.FORM,
+      tab: <TabIcon title="Form" iconFontName={faListUl} />,
```
I'm not sold on the `Form` title.
I'd be inclined to go with the opposite operation to the other tab: `encode`

alfetopito reviewed (approved):
Couple of nitpicks regarding the code but nothing major.
All of them can be addressed in follow up PRs.

I'm happy merging as is and iterate.

As I mentioned today, this page will be hidden until the new menu is ready, so it's fine to be merged even if some things still need to be adjusted.

I also mentioned to the team some more information to be displayed in each of the new types to give more context to users.

matextrem commented:
@alfetopito fixed! 

Cc: @elena-zh

elena-zh commented:
Hey @matextrem , could toy please describe what exactly I need to retest except changing the page path to /appdata?

alfetopito commented on src/const.ts:
```diff
@@ -245,9 +245,7 @@ export const COW_SDK_DEV = [Network.MAINNET, Network.RINKEBY, Network.GNOSIS_CHA
-    acc[networkId] = new CowSdk(networkId, {
-      isDevEnvironment: true,
-    })
+    acc[networkId] = new CowSdk(networkId)
```
Any reason for changing this?

I haven't checked it throughly, but I suspect it might be the cause of duplicated orders, like seen [in this batch](https://pr168--explorer.review.gnosisdev.com/tx/0x5c31c2b6d7ea91d67bb1c2baa589b5b0e861461896e20928b746056b3380451c/?tab=orders)

I would avoid touching this if possible and leave that up to Ramiro's/mine PR which is adding Goerli support and also updating the SDK that does this
![Screen Shot 2022-08-18 at 11 29 38](https://user-images.githubusercontent.com/43217/185374311-81a170fe-1eec-4fb9-a08b-c3ddba4bbc1b.png)

matextrem replied:
I think this broke when I merged from develop. Will fix this

matextrem commented:
> Hey @matextrem , could toy please describe what exactly I need to retest except changing the page path to /appdata and changing the tab name to 'Encode'?

@elena-zh I made some changes in the logic for the encode tab, so maybe we could test it again to see nothing is broken

elena-zh commented:
Seems like works as expected. 
Anyways, I will completely retest this page when it is going to be released. 
Thanks!
